CUDA: Optimize `reduce_rows_f32` kernel, leading up to 25x perf improvement on kernel-level and 10% perf increase for Gemma3n

ORippler opened the pull request:
Investigation of Gemma3n perf on NVGPUs identified the `reduce_rows_f32` kernel as a major performance bottleneck. Profiling revealed the kernel to be severely latency-limited in the regime run by by Gemma3n (nrows ~10, ncols in `[2048, 8192]`).

This PR addresses this issue, hiding the latency by a combination of:
1. Manual loop unrolling, getting the compiler to request all unrolled datapoints at once, instead of fetching data sequentially (`#pragma unroll` did not do the trick unfortunately).
2. Increasing the number of threads processing a row, where 512 threads are used for the low-parallelization regime (i.e. processing only a single row). This gives the SM 16 full warps to cycle through, further pipelining data fetching.

Since perf regressions were identified in the high-parallelization regime (nrows >= 2x SM count), we use:
* 128 threads for medium-to-large columns, effectively letting each SM process a single row (a SM can execute 4 warps x 32 threads=128 threads concurrently).
* As perf regression were still observed for small columns (< 1024 cols = 1 unrollment loop of a threadblock with size 128 and 8 unrolls), thread count was reduced to 32 threads for small columns. An alternative to this would have been to template the number of unrolls based on the column size. However, this would lead to an increased binary size due to the required compilation of multiple kernels, and was thus not pursued further.

The high/low parallelization threshold was empirically determined:
| GPU Model                     | Nrow SM Count Multiple, where 128 beats 512 threads |
| -----------                   | -----------            |
| RTX 4000 SFF ADA              | 2.0x                   |
| RTX 6000 ADA                  | 2.5x                   |
| RTX PRO 6000 Blackwell Max-Q  | 3.04x                  |
| RTX PRO 4500 Blackwell        | 3.15x                  |

In total, up to ~25x perf improvement was observed on kernel-level.
<img width="1804" height="574" alt="speedup_comparison_multiple" src="https://github.com/user-attachments/assets/3a9cbbda-eac8-408f-b00a-f0791abc10f0" />
Moreover, regression was not observed in any of the investigated combinations.
<img width="1795" height="574" alt="speedup_comparison_fractional" src="https://github.com/user-attachments/assets/4a8646bb-641b-4eeb-95e3-37385c068293" />

As a consequence of this general kernel optimization, Gemma3n achieves ~10% perf increase, going from 130 to 145 tok/s on a RTX PRO 6000 Blackwell-Max-Q with batch-size 1.

Naive:
```
  Device 0: NVIDIA RTX PRO 6000 Blackwell Max-Q Workstation Edition, compute capability 12.0, VMM: yes
| model                          |       size |     params | backend    | ngl | n_batch |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | --------------: | -------------------: |
| gemma3n E2B Q8_0               |   4.45 GiB |     4.46 B | CUDA       |  99 |       1 |           pp100 |        147.27 ± 0.82 |
| gemma3n E2B Q8_0               |   4.45 GiB |     4.46 B | CUDA       |  99 |       1 |           tg100 |        130.75 ± 0.28 |
```
Optimized
```
  Device 0: NVIDIA RTX PRO 6000 Blackwell Max-Q Workstation Edition, compute capability 12.0, VMM: yes
| model                          |       size |     params | backend    | ngl | n_batch |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | --------------: | -------------------: |
| gemma3n E2B Q8_0               |   4.45 GiB |     4.46 B | CUDA       |  99 |       1 |           pp100 |        168.41 ± 0.37 |
| gemma3n E2B Q8_0               |   4.45 GiB |     4.46 B | CUDA       |  99 |       1 |           tg100 |        146.68 ± 0.68 |
```
---
Side note: Similar tendencies were observed for `rms_norm_f32`, and we intend to optimize said kernel in a separate PR.

ORippler committed:
Factor out `reduce_rows_f32` from common.cuh

This increases iteration cycle speed by not having to recompile
every kernel all the time

ORippler committed:
Hide memory-latency by loop unrolling in reduce_rows_f32

ORippler committed:
Further optimizations to `reduce_rows_f32`

1. Increase threadblock size to better hide latency of memory requests.
   As a consequence of bigger threadblocks, do 2-step summation, using
   shared memory to communicate results between invocations
2. Use sum_temp array to reduce waits on sum
3. Adjust num_unroll to reflext bigger threadblock
4. Improve default block_dims, increase support for more block_dims

ORippler committed:
Add perf tests for `reduce_rows_f32` kernel

ORippler committed:
Add heuristic to toggle 128/512 threads based on sm count

Break even point was the minimum of the following multiples.

| GPU Model                     | Nrow SM Count Multiple |
| -----------                   | -----------            |
| RTX 4000 SFF ADA              | 2.0x                   |
| RTX 6000 ADA                  | 2.5x                   |
| RTX PRO 6000 Blackwell Max-Q  | 3.04x                  |
| RTX PRO 4500 Blackwell	| 3.15x                  |

ORippler committed:
Ensure perf gains also for small ncols and large nrows

Alternative to this, one could have also made the number of unrollings
template-able, but that would require compiling the kernel multiple
times, increasing binary size unnecessarily

ORippler committed:
Modify perf and unit-tests

ORippler committed:
Apply auto-formatting by clang

ORippler commented:
Rebased on current master, resolving conflicts along the way. Reran E2E perf tests for gemma3n, and nice to see some optimizations for `tg` were made apparently 😃 :


Naive: 
  Device 0: NVIDIA RTX PRO 6000 Blackwell Max-Q Workstation Edition, compute capability 12.0, VMM: yes
| model                          |       size |     params | backend    | ngl | n_batch |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | --------------: | -------------------: |
| gemma3n E2B Q8_0               |   4.45 GiB |     4.46 B | CUDA       |  99 |       1 |           pp100 |        146.89 ± 0.12 |
| gemma3n E2B Q8_0               |   4.45 GiB |     4.46 B | CUDA       |  99 |       1 |           tg100 |        145.86 ± 0.13 |

Optimized:
  Device 0: NVIDIA RTX PRO 6000 Blackwell Max-Q Workstation Edition, compute capability 12.0, VMM: yes
| model                          |       size |     params | backend    | ngl | n_batch |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | --------------: | -------------------: |
| gemma3n E2B Q8_0               |   4.45 GiB |     4.46 B | CUDA       |  99 |       1 |           pp100 |        167.47 ± 0.29 |
| gemma3n E2B Q8_0               |   4.45 GiB |     4.46 B | CUDA       |  99 |       1 |           tg100 |        167.28 ± 0.35 |

ORippler committed:
Fix CI build failure

See https://github.com/ggml-org/llama.cpp/actions/runs/16798370266/job/47573716079?pr=15132#step:7:486
Building with VS generator worked though.

JohannesGaessler commented on ggml/src/ggml-cuda/common.cuh:
```diff
@@ -864,6 +844,7 @@ struct ggml_cuda_graph {
 struct ggml_backend_cuda_context {
     int device;
     std::string name;
+    int sm_count;
```
Instead of adding a property with duplicated information, please retrieve it as is done e.g. [here](https://github.com/ggml-org/llama.cpp/blob/1d72c841888b9450916bdd5a9b3274da380f5b36/ggml/src/ggml-cuda/mmq.cuh#L3544-L3547).

ORippler replied:
> Instead of adding a property with duplicated information, please retrieve it as is done e.g. [here](https://github.com/ggml-org/llama.cpp/blob/1d72c841888b9450916bdd5a9b3274da380f5b36/ggml/src/ggml-cuda/mmq.cuh#L3544-L3547).

Thanks, was unaware of this design pattern of ggml.

JohannesGaessler commented on ggml/src/ggml-cuda/reduce_rows.cuh:
```diff
@@ -0,0 +1,53 @@
+    float     sum_temp[num_unroll] = { 0.0f };
+    for (int i = col; i < ncols;) {
+        for (int j = 0; j < num_unroll; ++j) {
+            if (i < ncols) {
```
My intuition would have been that it is faster not to add the inner loop due to this conditional statement. Just to be sure: did you test both versions?

ORippler replied:
> My intuition would have been that it is faster not to add the inner loop due to this conditional statement. Just to be sure: did you test both versions?

We shared that intuition and, as mentioned in the PR description, one of the first things we tried was hinting the compiler to unroll the outer loop with `#pragma unroll`. Unfortunately, the compiler did not comply, and we were still seeing a lot of long scoreboard stalls caused by sequential iteration through the for loop (see the following two screenshots).

<img width="2721" height="1246" alt="Screenshot 2025-08-07 at 13 28 59" src="https://github.com/user-attachments/assets/2b31886b-1404-4af1-aabc-7b6948adf4d4" />
<img width="2728" height="1237" alt="image" src="https://github.com/user-attachments/assets/05d9ad54-8689-4bb1-8454-0e69d30075de" />

Only by explicitly unrolling the loop did we get the compiler to comply and pre-fetch the data, effectively hiding the memory-latency (see 8 sequential `FADD`s preceeded by 8 sequential `LDG`s):
<img width="2723" height="1235" alt="Screenshot 2025-08-07 at 13 28 39" src="https://github.com/user-attachments/assets/fd9d3da8-c261-420d-b909-9d3de4bb27ec" />

JohannesGaessler commented on ggml/src/ggml-cuda/reduce_rows.cuh:
I would have written this kernel differently. I would have made the CUDA block size a template parameter and increased it as long as it reduces the number of iterations needed (as is done in e.g. `mmv.cu`/`mmvf.cu`).

JohannesGaessler commented on ggml/src/ggml-cuda/mean.cu:
Is Gemma3n being bottlenecked by `GGML_SUM` or `GGML_MEAN`? The reason I'm asking is because `GGML_SUM` uses CUB while `GGML_MEAN` does not. I would welcome a better general kernel for reducing rows in ggml but I would assume that such a kernel would not be faster than CUB.

slaren replied:
If you are interested in optimizing this model further, fusing these operations is likely to result in much better performance than optimizing the individual operations:

https://github.com/ggml-org/llama.cpp/blob/9a96389544a08fd829fccda28142ce2066017fde/src/llama-model.cpp#L10729-L10731

https://github.com/ggml-org/llama.cpp/blob/9a96389544a08fd829fccda28142ce2066017fde/src/llama-model.cpp#L10799-L10807

Fused operations in the CUDA backend are handled here:
https://github.com/ggml-org/llama.cpp/blob/9a96389544a08fd829fccda28142ce2066017fde/ggml/src/ggml-cuda/ggml-cuda.cu#L2886-L2899

ORippler committed:
Remove sm_count property from `ggml_backend_cuda_context`

Requested by @JohannesGaessler, and should fix remaining CI issues as a
side-effect

JohannesGaessler commented:
Thank you for answering my questions (even though I could have gotten the answers by reading the PR description more carefully). If you test using CUB for `GGML_MEAN` this PR would essentially be good to merge from my side.

ORippler committed:
Add CUB-based implementation for GGML_OP_MEAN

Currently this branch is only executed for nrows==1

ORippler committed:
Add heuristics to execute CUB branch only when it brings perf

Heuristics were determined on the following HW:

* RTX 4000 SFF ADA
* RTX 6000 ADA
* RTX PRO 6000 Blackwell Max-Q
* RTX PRO 4500 Blackwell

ORippler committed:
Add unit-test for CUB-based mean

Tests should run with CUDA Graphs enabled per default on NVGPUs

JohannesGaessler commented on ggml/src/ggml-cuda/common.cuh:
```diff
@@ -87,6 +87,10 @@
 #define GGML_CUDA_CC_IS_NG(cc)       (cc >= GGML_CUDA_CC_NG)
 
+#if !defined(GGML_USE_HIP) && !defined(GGML_USE_MUSA) && CUDART_VERSION >= 11070
+#    define USE_CUB
```
```suggestion
#define GGML_CUDA_USE_CUB
```

Since you're moving the macro into the common header, please use this prefix.

JohannesGaessler commented on ggml/src/ggml-cuda/mean.cu:
```diff
@@ -1,4 +1,14 @@
+#include "reduce_rows.cuh"
+
+#ifdef USE_CUB
+#    include <cub/cub.cuh>
```
```suggestion
#include <cub/cub.cuh>
```

ORippler replied:
Our [`.clang-format`](https://github.com/ggml-org/llama.cpp/blob/master/.clang-format#L92) specifies to indent preprocessing directives after the hash, so I kept it as is. Let me know should the path forward not be to apply the specified formatting rules on an as-we-go basis to consolidate the code formatting.

JohannesGaessler replied:
Preferably keep consistency with the surrounding code, where I am definitely not indenting defines.

IMbackK replied:
Probably something that should be addressed and something i have mentioned [before](https://github.com/ggml-org/llama.cpp/pull/11362#issuecomment-2613334540). Greping this repo shows that all three possibilites are in use:

1.
```
int func()
{
#    define foo
}
```

2.
```
int func()
{
#define foo
}
```

3.
```
int func()
{
    #define foo
}
```

i really dont like 1. but 2. and 3. are fine with me personally

IMbackK replied:
https://github.com/ggml-org/llama.cpp/issues/15264

ORippler replied:
> Preferably keep consistency with the surrounding code, where I am definitely not indenting defines.

All right, then I'll unindent them.

JohannesGaessler commented on ggml/src/ggml-cuda/mean.cu:
```diff
@@ -13,7 +23,45 @@ void ggml_cuda_op_mean(ggml_backend_cuda_context & ctx, ggml_tensor * dst) {
+        divide_by_count<float><<<1, 1, 0, stream>>>(dst_d, ncols);
+        return;
+    }
+#endif
```
```suggestion
#endif // GGML_CUDA_USE_CUB
```

JohannesGaessler reviewed (approved):
Thank you for the high-effort PR.

ORippler committed:
Rename `USE_CUB` to `GGML_CUDA_USE_CUB`

Suggested by @JohannesGaessler

ORippler committed:
Unindent Preprocessor directives

See
https://github.com/ggml-org/llama.cpp/pull/15132#discussion_r2269213506

ORippler commented:
@JohannesGaessler Could we get this merged whenever you have the time? Unfortunately I don't have write access 🙈

ggerganov commented:
This change breaks the builds with `GGML_CUDA_GRAPHS=OFF`:

https://github.com/ggml-org/ci/blob/results/ggml/56/06bd129afb75630a6fe09552cdc37dc73d636d/ggml-4-x86-cuda-v100/stdall#L248
